File: backend/app/models/token_distribution.py
from decimal import Decimal
from typing import Optional
from sqlalchemy import String, Numeric, Integer, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base
from .token_configuration import TokenConfiguration


class TokenDistribution(Base):
    """Model for tracking token distributions to recipients"""
    
    __tablename__ = "token_distributions"

    # Distribution details
    token_config_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("token_configurations.id", ondelete="CASCADE"),
        nullable=False
    )
    recipient_address: Mapped[str] = mapped_column(String, nullable=False)
    chain_id: Mapped[str] = mapped_column(String, nullable=False)
    token_amount: Mapped[Decimal] = mapped_column(Numeric, nullable=False)
    status: Mapped[str] = mapped_column(
        String,
        nullable=True,
        server_default="pending"
    )
    transaction_hash: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    # Relationship
    token_config: Mapped[TokenConfiguration] = relationship(
        "TokenConfiguration",
        back_populates="distributions"
    ) 

File: backend/src/models/TokenDistribution.js
const { DataTypes } = require('sequelize');
const sequelize = require('../db/config');
const TokenConfiguration = require('./TokenConfiguration');

const TokenDistribution = sequelize.define('TokenDistribution', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  tokenConfigId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    field: 'token_config_id',
    references: {
      model: 'token_configurations',
      key: 'id'
    }
  },
  recipientAddress: {
    type: DataTypes.STRING,
    allowNull: false,
    field: 'recipient_address'
  },
  chainId: {
    type: DataTypes.STRING,
    allowNull: false,
    field: 'chain_id'
  },
  tokenAmount: {
    type: DataTypes.DECIMAL,
    allowNull: false,
    field: 'token_amount'
  },
  status: {
    type: DataTypes.STRING,
    allowNull: false,
    defaultValue: 'pending'
  },
  transactionHash: {
    type: DataTypes.STRING,
    field: 'transaction_hash'
  }
}, {
  tableName: 'token_distributions',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at'
});

// Establish relationship
TokenDistribution.belongsTo(TokenConfiguration, { foreignKey: 'tokenConfigId' });
TokenConfiguration.hasMany(TokenDistribution, { foreignKey: 'tokenConfigId' });

module.exports = TokenDistribution; 